File: 01-sistema-gestion-datos/dominio/interfaces/repositorio_usuario.py
"""
Interface RepositorioUsuario - Capa de Dominio
Define el contrato para el repositorio de usuarios
"""
from abc import ABC, abstractmethod
from typing import List, Optional
from ..entidades.usuario import Usuario
from ..value_objects.email import Email
from ..value_objects.nombre_usuario import NombreUsuario


class RepositorioUsuario(ABC):
    """
    Interface que define el contrato para el repositorio de usuarios
    Implementa el patrón Repository para abstraer el acceso a datos
    """
    
    @abstractmethod
    async def guardar(self, usuario: Usuario) -> Usuario:
        """
        Guardar un usuario en el repositorio
        
        Args:
            usuario: Usuario a guardar
            
        Returns:
            Usuario guardado con ID asignado
            
        Raises:
            ErrorRepositorio: Si hay error al guardar
        """
        pass
    
    @abstractmethod
    async def obtener_por_id(self, id_usuario: int) -> Optional[Usuario]:
        """
        Obtener usuario por ID
        
        Args:
            id_usuario: ID del usuario
            
        Returns:
            Usuario encontrado o None si no existe
            
        Raises:
            ErrorRepositorio: Si hay error al consultar
        """
        pass
    
    @abstractmethod
    async def obtener_por_email(self, email: Email) -> Optional[Usuario]:
        """
        Obtener usuario por email
        
        Args:
            email: Email del usuario
            
        Returns:
            Usuario encontrado o None si no existe
            
        Raises:
            ErrorRepositorio: Si hay error al consultar
        """
        pass
    
    @abstractmethod
    async def obtener_por_nombre_usuario(self, nombre_usuario: NombreUsuario) -> Optional[Usuario]:
        """
        Obtener usuario por nombre de usuario
        
        Args:
            nombre_usuario: Nombre de usuario
            
        Returns:
            Usuario encontrado o None si no existe
            
        Raises:
            ErrorRepositorio: Si hay error al consultar
        """
        pass
    
    @abstractmethod
    async def listar_todos(self, limite: int = 100, offset: int = 0) -> List[Usuario]:
        """
        Listar todos los usuarios con paginación
        
        Args:
            limite: Número máximo de usuarios a retornar
            offset: Número de usuarios a saltar
            
        Returns:
            Lista de usuarios
            
        Raises:
            ErrorRepositorio: Si hay error al consultar
        """
        pass
    
    @abstractmethod
    async def listar_activos(self, limite: int = 100, offset: int = 0) -> List[Usuario]:
        """
        Listar usuarios activos con paginación
        
        Args:
            limite: Número máximo de usuarios a retornar
            offset: Número de usuarios a saltar
            
        Returns:
            Lista de usuarios activos
            
        Raises:
            ErrorRepositorio: Si hay error al consultar
        """
        pass
    
    @abstractmethod
    async def actualizar(self, usuario: Usuario) -> Usuario:
        """
        Actualizar un usuario existente
        
        Args:
            usuario: Usuario con datos actualizados
            
        Returns:
            Usuario actualizado
            
        Raises:
            ErrorRepositorio: Si hay error al actualizar
            UsuarioNoEncontrado: Si el usuario no existe
        """
        pass
    
    @abstractmethod
    async def eliminar(self, id_usuario: int) -> bool:
        """
        Eliminar un usuario por ID
        
        Args:
            id_usuario: ID del usuario a eliminar
            
        Returns:
            True si se eliminó correctamente, False si no existía
            
        Raises:
            ErrorRepositorio: Si hay error al eliminar
        """
        pass
    
    @abstractmethod
    async def existe_email(self, email: Email) -> bool:
        """
        Verificar si existe un usuario con el email dado
        
        Args:
            email: Email a verificar
            
        Returns:
            True si existe, False si no
            
        Raises:
            ErrorRepositorio: Si hay error al consultar
        """
        pass
    
    @abstractmethod
    async def existe_nombre_usuario(self, nombre_usuario: NombreUsuario) -> bool:
        """
        Verificar si existe un usuario con el nombre de usuario dado
        
        Args:
            nombre_usuario: Nombre de usuario a verificar
            
        Returns:
            True si existe, False si no
            
        Raises:
            ErrorRepositorio: Si hay error al consultar
        """
        pass
    
    @abstractmethod
    async def contar_usuarios(self) -> int:
        """
        Contar el número total de usuarios
        
        Returns:
            Número total de usuarios
            
        Raises:
            ErrorRepositorio: Si hay error al consultar
        """
        pass

File: 06-arquitectura-microservicios-java/microservicios/servicio-usuarios/src/main/java/com/portfolio/arquitectura/usuarios/dominio/repositorio/RepositorioUsuario.java
package com.portfolio.arquitectura.usuarios.dominio.repositorio;

import com.portfolio.arquitectura.usuarios.dominio.entidad.Usuario;
import com.portfolio.arquitectura.usuarios.dominio.entidad.RolUsuario;
import com.portfolio.arquitectura.usuarios.dominio.entidad.EstadoUsuario;
import com.portfolio.arquitectura.usuarios.dominio.objetoValor.CorreoElectronico;

import java.util.List;
import java.util.Optional;
import java.util.UUID;

/**
 * Interfaz del Repositorio de Usuario - Capa de Dominio
 * 
 * Define el contrato para la persistencia de usuarios.
 * Implementa el patrón Repository para desacoplar la lógica de negocio
 * de la implementación de persistencia.
 * 
 * @author Gabriel - Arquitecto de Software
 */
public interface RepositorioUsuario {

    /**
     * Guarda un usuario en el repositorio
     * 
     * @param usuario Usuario a guardar
     * @return Usuario guardado con ID generado
     */
    Usuario guardar(Usuario usuario);

    /**
     * Busca un usuario por ID
     * 
     * @param id ID del usuario
     * @return Optional con el usuario si existe
     */
    Optional<Usuario> buscarPorId(UUID id);

    /**
     * Busca un usuario por correo electrónico
     * 
     * @param correo Correo electrónico del usuario
     * @return Optional con el usuario si existe
     */
    Optional<Usuario> buscarPorCorreo(CorreoElectronico correo);

    /**
     * Busca un usuario por nombre de usuario
     * 
     * @param nombreUsuario Nombre de usuario
     * @return Optional con el usuario si existe
     */
    Optional<Usuario> buscarPorNombreUsuario(String nombreUsuario);

    /**
     * Verifica si existe un usuario con el correo dado
     * 
     * @param correo Correo electrónico a verificar
     * @return true si existe un usuario con ese correo
     */
    boolean existePorCorreo(CorreoElectronico correo);

    /**
     * Verifica si existe un usuario con el nombre de usuario dado
     * 
     * @param nombreUsuario Nombre de usuario a verificar
     * @return true si existe un usuario con ese nombre de usuario
     */
    boolean existePorNombreUsuario(String nombreUsuario);

    /**
     * Busca usuarios por estado
     * 
     * @param estado Estado del usuario
     * @return Lista de usuarios con el estado especificado
     */
    List<Usuario> buscarPorEstado(EstadoUsuario estado);

    /**
     * Busca usuarios por rol
     * 
     * @param rol Rol del usuario
     * @return Lista de usuarios con el rol especificado
     */
    List<Usuario> buscarPorRol(RolUsuario rol);

    /**
     * Busca usuarios con paginación
     * 
     * @param pagina Número de página (0-based)
     * @param tamano Tamaño de la página
     * @return Lista paginada de usuarios
     */
    List<Usuario> buscarTodos(int pagina, int tamano);

    /**
     * Cuenta el total de usuarios
     * 
     * @return Número total de usuarios
     */
    long contar();

    /**
     * Cuenta usuarios por estado
     * 
     * @param estado Estado del usuario
     * @return Número de usuarios con el estado especificado
     */
    long contarPorEstado(EstadoUsuario estado);

    /**
     * Cuenta usuarios por rol
     * 
     * @param rol Rol del usuario
     * @return Número de usuarios con el rol especificado
     */
    long contarPorRol(RolUsuario rol);

    /**
     * Elimina un usuario por ID
     * 
     * @param id ID del usuario a eliminar
     */
    void eliminarPorId(UUID id);

    /**
     * Busca usuarios por nombre (búsqueda parcial)
     * 
     * @param nombre Nombre a buscar
     * @param apellido Apellido a buscar
     * @return Lista de usuarios que coinciden
     */
    List<Usuario> buscarPorNombreConteniendoIgnoreCaseOrApellidoConteniendoIgnoreCase(
            String nombre, String apellido);

    /**
     * Busca usuarios activos con paginación
     * 
     * @param pagina Número de página (0-based)
     * @param tamano Tamaño de la página
     * @return Lista paginada de usuarios activos
     */
    List<Usuario> buscarUsuariosActivos(int pagina, int tamano);

    /**
     * Busca usuarios por rango de fechas de creación
     * 
     * @param fechaInicio Fecha de inicio
     * @param fechaFin Fecha de fin
     * @return Lista de usuarios creados en el rango
     */
    List<Usuario> buscarPorFechaCreacionEntre(java.time.LocalDateTime fechaInicio, 
                                             java.time.LocalDateTime fechaFin);
}
